internal/core/application/operator_service.go

```diff
@@ -1656,21 +1664,67 @@ func (o *operatorService) GetMarketReport(
 
 			volume[swapRequest.GetAssetR()] += int64(swapRequest.GetAmountR())
 			volume[swapRequest.GetAssetP()] += int64(swapRequest.GetAmountP())
+
+			for i, v := range groupedVolume {
+				//find time slot to which trade belongs to and calculate volume for that slot
+				if (time.Unix(int64(trade.SwapRequest.Timestamp), 0).After(v.StartTime) ||
+					time.Unix(int64(trade.SwapRequest.Timestamp), 0).Equal(v.StartTime)) &&
+					(time.Unix(int64(trade.SwapRequest.Timestamp), 0).Before(v.EndTime) ||
+						time.Unix(int64(trade.SwapRequest.Timestamp), 0).Equal(v.EndTime)) {
+
+					//assume AmountR is base asset, AmountP(FeeAsset) is quote asset
+					volumeBaseAmount := swapRequest.GetAmountR() + v.BaseVolume
+					volumeQuoteAmount := swapRequest.GetAmountP() + v.QuoteVolume
+					if swapRequest.GetAssetR() == market.QuoteAsset {
+						volumeBaseAmount = swapRequest.GetAmountP() + v.BaseVolume
+						volumeQuoteAmount = swapRequest.GetAmountR() + v.QuoteVolume
+					}
+
+					groupedVolume[i] = MarketVolume{
+						BaseVolume:  volumeBaseAmount,
+						QuoteVolume: volumeQuoteAmount,
+						StartTime:   v.StartTime,
+						EndTime:     v.EndTime,
+					}
+				}
+			}
 		}
 	}
 
 	return &MarketReport{
 		CollectedFees: MarketCollectedFees{
 			BaseAmount:  uint64(totalFees[market.BaseAsset]),
 			QuoteAmount: uint64(totalFees[market.QuoteAsset]),
+			StartTime:   startTime,
+			EndTime:     endTime,
 		},
 		Volume: MarketVolume{
 			BaseVolume:  uint64(volume[market.BaseAsset]),
-			QuoteVolume: uint64(volume[market.BaseAsset]),
+			QuoteVolume: uint64(volume[market.QuoteAsset]),
+			StartTime:   startTime,
+			EndTime:     endTime,
 		},
+		GroupedVolume: groupedVolume,
 	}, nil
 }
 
+// initGroupedVolume initialised list of MarketVolume with grouped time slots(start/end date)
+//ordered desc
+func initGroupedVolume(start, end time.Time, groupByHours int) []MarketVolume {
+	groupedVolume := make([]MarketVolume, 0)
+	for {
+		if end.Equal(start) || end.Before(start) {
+			return groupedVolume
+		} else {
+			groupedVolume = append(groupedVolume, MarketVolume{
+				StartTime: end.Add(-time.Hour * time.Duration(groupByHours)),
+				EndTime:   end,
+			})
+			end = end.Add(-time.Hour * time.Duration(groupByHours))
+		}
+	}
```

Comment:
we must add here just a second more to avoid collisions between `MarketVolume[i].Start` and `MarketVolume[i+1].End`, which would cause, then, a trade with exactly that timestamp to be repeated in 2 different slots.
The very last slot doesn't need this extra second, but its start time must be exactly the timerange's start time:
```suggestion
		} else {
		        nextEnd := end.Add(-time.Hour * time.Duration(groupByHours))
		        nextStart := start
		        if nextEnd.Sub(start).Seconds() > 0 {
		            nextStart = nextEnd.Add(time.Second)
		        }
			groupedVolume = append(groupedVolume, MarketVolume{
				StartTime: nextStart,
				EndTime:   end,
			})
			end = nextEnd
		}
	}
```

Reply:
Done